File: 240327/갈 수 있는 곳들/places-can-go.cpp
// https://www.codetree.ai/missions/2/problems/places-can-go/description
#include <iostream>
#include <queue>
using namespace std;

int n, k, cnt;	// cnt : 방문 가능한 칸 수
int dx[4] = { 1,0,-1,0 };
int dy[4] = { 0,1,0,-1 };
int grid[101][101];
bool visited[101][101];
queue<pair<int, int>> q;

// 출발점이 1-based 기준이라 range check도 1-based 방식으로 변경
bool InRange(int x, int y) {
	return x >= 1 && x <= n && y >= 1 && y <= n;
}

bool canGo(int x, int y) {
	if (!InRange(x, y))
		return false;
	// 방문 가능한 지점이 1이 아니라 0임
	else if (visited[x][y] || grid[x][y] == 1)
		return false;
	else
		return true;
}

int main() {
	cin >> n >> k;

	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			cin >> grid[i][j];
		}
	}

	int x, y;
	for (int i = 0; i < k; i++) {
		cin >> x >> y;

		// 이미 출발점을 방문한 경우 해당 위치에서 BFS 진행하지 않음
		if (visited[x][y])
			continue;

		visited[x][y] = true;
		q.push(make_pair(x, y));

		while (!q.empty()) {
			int cur_x = q.front().first;
			int cur_y = q.front().second;
			cnt++;	// 방문 가능한 지점 +1 추가
			q.pop();

			for (int i = 0; i < 4; i++) {
				int new_x = cur_x + dx[i];
				int new_y = cur_y + dy[i];

				if (canGo(new_x, new_y)) {
					visited[new_x][new_y] = true;
					q.push(make_pair(new_x, new_y));
				}
			}
		}
	}

	cout << cnt << '\n';
	return 0;
}

File: 240629/갈 수 있는 곳들/places-can-go.js
// 문제 https://www.codetree.ai/missions/2/problems/places-can-go/description

// 이동할 수 있는 곳 -> 0
// 이동할 수 없는 곳 -> 1

class Queue {
    constructor() {
        this._arr = [];
    }

    enqueue(item) {
        this._arr.push(item);
    }

    dequeue() {
        return this._arr.shift();
    }

    isEmpty() {
        return this._arr.length === 0 ? true : false;
    }
}

let N = null;
let k = null;
let countN = 0;
let countK = 0;

let dx = [-1, 0, 1, 0];
let dy = [0, 1, 0, -1];

let graph = [];
let start = [];

const solution = (graph, visited, start) => {
    const inRange = (x, y) => {
        return x >= 0 && x < N && y >= 0 && y < N;
    }

    const canGo = (x, y) => {
        if (!inRange(x, y)) return false;
        else if (visited[x][y] || graph[x][y]) return false;
        else return true;
    }

    for (let [x, y] of start) {
        const queue = new Queue();

        // 0-base graph이므로 좌표에서 1씩 빼주기
        queue.enqueue([x - 1, y - 1]);
        visited[x - 1][y - 1] = true;

        while (!queue.isEmpty()) {
            let [curX, curY] = queue.dequeue();

            for (let i = 0; i < 4; i++) {
                let newX = curX + dx[i];
                let newY = curY + dy[i];

                if (canGo(newX, newY)) {
                    queue.enqueue([newX, newY]);
                    visited[newX][newY] = true;
                }
            }
        }
    }

    // 방문한 노드의 개수를 세서 저장하는 과정
    let res = visited.reduce((prev, curr) => prev + curr.filter(e => e).length, 0);
    console.log(res);
}

const rl = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout,
});

rl.on('line', (line) => {
    if (!N && !k) {
        [N, k] = line.split(' ').map(e => +e);
    }
    else if (countN < N) {
        graph.push(line.split(' ').map(e => +e));
        countN++;
    }
    else {
        start.push(line.split(' ').map(e => +e));
        countK++;
    }

    if (countK === k) {
        rl.close();
    }
}).on('close', () => {
    let visited = Array.from(Array(N), () => Array(N).fill(false));
    solution(graph, visited, start);
    process.exit();
});